### PR TITLE
Allows dependency version comparison to use up to "patch" number

### LINKED ISSUE #5356: Dependencies version check not working for modules that use three digits (e.g. 1.0.14)
**Description of the bug**

The dependencies version check specified in `info` files doesn't work properly when checking the version of a module that uses 3 digits for versioning, such as `entity_plus (>1.0.14)`. 

**Steps To Reproduce**

1. Create a custom module. In the info file, specify a dependency from a specific version of another module that uses three digits to define its version. For example, `dependencies[] = entity_plus (>1.0.14)`. 
2. Try to enable the module. Backdrop should not allow this module to be enable (for this example, the latest available version of Entity Plus is 1.0.13). 

**Actual behavior**
Backdrop allows you to enable the custom module.

**Expected behavior**
Backdrop should not allow the custom module to be enabled, since, in this example, it depends on a version of Entity Plus that is not yet available.   (Note: this issue happens regardless of the module, as long as its version uses 3 digits instead of 2).

**Additional information**

It seems like the version check performed by core's function `backdrop_parse_dependency()` is only extracting the first two digits of the version, and ignores the third digit. For example, using Devel, run the following:

```php
dpm(backdrop_parse_dependency('entity_plus (>1.0.14)'));
```

You will obtain the attached screenshot. The important array key here is `version`, whose value is `1.0` instead of `1.0.14`, resulting on a failed check.

I don't have enough knowledge of RegExp to rewrite the pattern to output the version correctly.  

Screen Shot 2021-11-17 at 10 27 17 AM

### PR DESCRIPTION
**Pull request by argiepiano**
Fixes https://github.com/backdrop/backdrop-issues/issues/5356

This patch increases the granularity of dependency checks in `info` files to compare dependencies using up to three numbers in versions (`major`.`minor`.`patch`). 

For example you may specify:

`dependencies[] = entity_plus (>=1.0.13)` 

The line above will prevent a module from being enabled unless the site has installed entity_plus version 1.0.13 or higher. Currently, Backdrop only checks `major`.`minor` version. This means that the above line would allow the module to be enabled with ANY version of entity_plus that started with 1.0  (e.g. 1.0.10) 

Additionally, this patch allows the use of "x" both in `minor` and `patch` as a wildcard to allow any number. Thus:
* `=1.x.x` is true for major version 1, and any minor and patch versions of a module 
* `>=1.x.x` is true for  any version 1 or higher
* `>1.x.x` is true for any version 2 or higher, and is the equivalent to `>1` or `>=2`
* Etc.

I believe I have tested most permutations, but please do additional checks on your end.

### CODE REVIEW

**Comment by argiepiano** [comment]
One potential caveat is the following: you may specify `dependencies[] = my_module (>1.2)` assuming that this will disallow enabling a module unless version `1.3` or higher of my_module installed. However, having version `1.2.1` of my_module will allow enabling, as `version_compare('1.2.1', '1.2', '>')` returns true .    

To prevent this you need to specify either `>1.2.x` or `>=1.3`

**Comment by herbdool** [comment]
@argiepiano I'm ok with that caveat. It seems that it's almost always smarter to specify >= rather than >.

**Comment by argiepiano** [comment]
I'm adding a commit with a modified the version dependency check tests. Questions:

- For comparisons with a RELEASED module (in the test the module is called `common_test.module`) I'm using a comparison against a  "normal" release pattern consisting of `major.minor.patch[-extra]`. In the test, version of `common_test.module` is 1.x-2.4.5-beta3.  
- For comparisons with DEV versions of a module, I kept the original dev pattern present in the original test: `1.x-2.x-dev` for the test module `common_test.module`. I believe that modules in dev stages do not use a "patch" portion - only the "2.x-dev" to indicate that this is a dev version

If this is not correct, please let me know.

**Comment by herbdool** [comment]
Sounds fine @argiepiano. I don't think anyone is creating dev releases anyway since it's easy to just download from GitHub if someone really wants that.

File: core/modules/system/tests/system.test [hunk 1]
```diff
@@ -561,42 +561,42 @@ class ModuleVersionTestCase extends ModuleTestCase {
...
+      'common_test (>2.x.x)' => FALSE,
       // Another branch compatibility.
-      'common_test (<=2.x)' => TRUE,
+      'common_test (<=2.x.x)' => TRUE,
       // Another branch incompatibility.
-      'common_test (<2.x)' => FALSE,
+      'common_test (<2.x.x)' => FALSE,
       // Another branch compatibility.
-      'common_test (>=2.x)' => TRUE,
+      'common_test (>=2.x.x)' => TRUE,
```
Comment (quicksketch) [inline core/modules/system/tests/system.test]: I think all these checks above should probably be reverted. A module specifying a dependency would only declare `>=2.x`. And the subsequent patch `.x` would be implied. You'd never see a `>=2.x.1` for example.

File: core/modules/system/tests/system.test [hunk 2]
```diff
@@ -561,42 +561,42 @@ class ModuleVersionTestCase extends ModuleTestCase {
...
+      'common_test (=1.x2.x.5, >=2.4.5)' => FALSE,
       // Core version is optional. Compatible.
-      'common_test (=1.x-2.x, >=2.4-alpha2)' => TRUE,
+      'common_test (=1.x-2.x.x, >=2.4.5-alpha2)' => TRUE,
       // Test !=, explicitly incompatible.
-      'common_test (=2.x, !=2.4-beta3)' => FALSE,
+      'common_test (=2.x, !=2.4.5-beta3)' => FALSE,
       // Three operations. Compatible.
-      'common_test (=2.x, !=2.3, <2.5)' => TRUE,
+      'common_test (=2.x.x, !=2.3.4, <2.5.0)' => TRUE,
```
Comment (quicksketch) [inline core/modules/system/tests/system.test]: Same as above, if doing a major version comparison, no patch level is needed.

```suggestion
      'common_test (=2.x, !=2.3.4, <2.5.0)' => TRUE,
```

File: core/modules/system/tests/system.test [hunk 3]
```diff
@@ -561,42 +561,42 @@ class ModuleVersionTestCase extends ModuleTestCase {
...
+      'common_test (>=1.x-2.4.5-beta2)' => TRUE,
       // Lower alpha version. Compatible.
-      'common_test (>=1.x-2.4-alpha8)' => TRUE,
+      'common_test (>=1.x-2.4.5-alpha8)' => TRUE,
       // Nonsense, misses a dash. Incompatible with everything.
-      'common_test (=1.x2.x, >=2.4)' => FALSE,
+      'common_test (=1.x2.x.5, >=2.4.5)' => FALSE,
       // Core version is optional. Compatible.
-      'common_test (=1.x-2.x, >=2.4-alpha2)' => TRUE,
+      'common_test (=1.x-2.x.x, >=2.4.5-alpha2)' => TRUE,
```
Comment (quicksketch) [inline core/modules/system/tests/system.test]:
```suggestion
      'common_test (=1.x-2.x, >=2.4.5-alpha2)' => TRUE,
```

**Review by quicksketch — COMMENTED** [review]
Thanks for adding tests @argiepiano! A few comments:

File: core/modules/system/tests/system.test (hunk 1 above)
Comment (argiepiano) [reply core/modules/system/tests/system.test]: That makes sense. I'll change those to your suggestion

**Comment by laryn** [comment]
Thank you @argiepiano, @herbdool, and @quicksketch -- merging into 1.x and 1.20.x.

**Pull request merged**